pytorch_lightning/trainer/supporters.py

```diff
@@ -90,3 +93,56 @@ def accumulate(self, x):
 
     def mean(self):
         return self.total / self.num_values
+
+
+class PredictionCollection(object):
+
+    def __init__(self, global_rank: int, world_size: int):
+        self.global_rank = global_rank
+        self.world_size = world_size
+        self.predictions = {}
+        self.num_predictions = 0
+
+    def _add_prediction(self, name, values, filename):
+        if filename not in self.predictions:
+            self.predictions[filename] = {name: values}
+        elif name not in self.predictions[filename]:
+            self.predictions[filename][name] = values
+        elif isinstance(values, Tensor):
+            self.predictions[filename][name] = torch.cat((self.predictions[filename][name], values.detach()))
+        elif isinstance(values, list):
+            self.predictions[filename][name].extend(values)
+
+    def add(self, predictions):
+
+        if predictions is None:
+            return
+
+        for filename, pred_dict in predictions.items():
+            for feature_name, values in pred_dict.items():
+                self._add_prediction(feature_name, values, filename)
+
+    def to_disk(self):
+        """Write predictions to file(s).
+        """
+        for filename, predictions in self.predictions.items():
+
+            # Filename or filename with rank extension in multi-gpu environment
+            outfile = Path(filename)
+            outfile = Path(f"{outfile.stem}{f'_rank_{self.global_rank}' if self.world_size > 1 else ''}{outfile.suffix}")
+            outfile.parent.mkdir(exist_ok=True, parents=True)
+
+            # Convert any tensor values to list
+            predictions = {k: v if not isinstance(v, Tensor) else v.tolist() for k, v in predictions.items()}
```

Comment:
if I predict segmentation maps here (or more general multidimensional tensors), will they be nested lists/will the shape be retained?

Reply:
This is really good feedback. I was assuming we would write to CSV. I could switch this to JSON and we could support any shape then. How does that sound?

Reply:
should be better I think. Otherwise we could also think of just dumping it with torch.save (basically pickle) or saving as a bumpy array

Reply:
went with torch.save 😄